src/lit-element.ts

```diff
@@ -58,59 +60,77 @@ export class LitElement extends UpdatingElement {
 
   /**
    * Array of styles to apply to the element. The styles should be defined
-   * using the `css` tag function.
+   * using the `css` tag function or via constructible stylesheets.
    */
-  static styles?: CSSResult|CSSResultArray;
+  static styles?: CSSResultOrNative|CSSResultArray;
 
   private static _styles: CSSResult[]|undefined;
+  private static _nativeStyles: CSSStyleSheet[]|undefined;
```

Comment:
Why does this need to be separate from `_styles`? Doesn't total ordering between the two sets matter too?

Reply:
We use one or the other depending on browser support. I also though about using a union TypeScript type but it got particularly awkward.

Reply:
well the concern about ordering/precedence still matters, right?

so to merge `_nativeStyles` into `_styles`, you'd want the type signature something like this:

```ts
private static _styles: (CSSResult | CSSStyleSheet)[] | undefined
```

then, when you're looping over each stylesheet, you'd distinguish whether it's a native stylesheet using the `instanceof` operator or something along those lines

Reply:
Adding `_nativeStyles` lets us assign at once to each element—we just set `.adoptedStyleSheets = this._nativeStyles`. I think it's much more ergonomic than casting every element of the array when you assign it.

I want to hear what @justinfagnani thinks—mostly I hope I've made it clear that we use _either_ array, so total ordering doesn't matter.

Reply:
> I think it's much more ergonomic than casting every element of the array when you assign it.

i'm not sure i follow.. casting every element when assigning to the array? maybe there's just a bit of confusion about the details of that type signature

the type signature `(CSSResult | CSSStyleSheet)[] | undefined` would allow the array to contain stylesheets of *either* type: so you'd be able to push both types of stylesheets to the array without any type casting

```ts
const cssResult: CSSResult = getExampleCssResult()
const nativeCss: CSSStyleSheet= getExampleNativeCss()

this._styles.push(cssResult) // no casting
this._styles.push(nativeCss) // no problem

for (const style of this._styles) {
  // style has type: (CSSResult | CSSStyleSheet)
}
```

however when you access one of the array values, you'd likely want to distinguish it's type to handle them separately

Reply:
The type signature should actually be `(CSSResult[] | CSSStyleSheet[])`, because you only have `CSSStyleSheet` where native adopted stylesheets are supported (regardless of input).

And then the only way you check the type is by doing something like `this._styles[0] instanceof CSSStyleSheet`, which is awkward.